src/posts/2020-09-19-how-to-write-accessible-forms.md

```diff
@@ -0,0 +1,154 @@
+
+The following principles apply for multi-step forms:
+
+- Repeat overall instruction on every page.
+- The forms should be split into logical groups.
+- The forms should inform the person about the progress thay are making.
+
+**Note**: There are no hard rules that defines a long form. However, you should consider if the form can be broken down into different logical sections.
+
+## Provide error messages
+
+When people encounter errors, they may need help to determine what's wrong and why. A common way forms indicate invalid fields is to change the `<input>` border color to red which is not ideal. The reason for this is because a change in color alone is not exposed to assistive technologies, so indicating errors in such ways will not inform the person of the error.
+
+The following are ways to effectively and accessibly communicate form errors:
+
+- Let the person know there is an error.
+- Let the person access the error, and provide instructions on how to fix it.
+- Allow the person re-submit the form.
+
+In the code below, there is an empty [live region](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) which report errors to assistive technologies. Additionally, each field is followed by a _hint_ which is hidden by default but shown if that field does not validate.
+
+```html
+<form>
+  <label for="username">Enter a username</label>
+  <input type="text" id="username">
+  <p class="hint hidden" id="usernameHint" >Your username cannot contain punctuation</p>
+
+  <label for="email">Enter your email address</label>
+  <input type="text" id="email">
+  <p class="hint hidden" id="emailHint">Please enter a valid email address</p>
+
+  <label for="password">Enter a password</label>
+  <input type="password" id="password">
+  <p class="hint hidden" id="passwordHint">Your password must be at least 6 characters</p>
+
+  <div aria-live="assertive" id="message"></div>
+ 
+  <button type="submit">Submit</button>
+</form>
+```
+
+```css
+.hidden {
+  display: none
+}
+
+.hint {
+  color: red;
+}
+```
+If validation errors occur, two things happen:
+
+1. The [live region](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) is populated upon form submission, announcing to assistive devices that there are errors.
+
+2. Each invalid `<input>` is given an [aria-invalid="true"](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-invalid_attribute) attribute and connected to a __hint__ using  an [aria-describedby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-describedby_attribute) attribute.
+
+Here's the markup for the email field in its invalid state:

```

Comment:
Add an empty line after this paragraph and before your code example so the markdown parser will render the code block.